### PR TITLE
primitives: deprecate SignalLogger and friends

### PR DESCRIPTION
**Pull request by rpoyner-tri**
Closes: #10228

This is step 4 of 4 to deprecate and replace SignalLogger and SignalLog
with a logging system that avoids threading hazards (SignalLogger) and
slow memory management (SignalLog).

The patch deprecates SignalLog, SignalLogger, and LogOutput, and
replaces them within Drake by VectorLog, VectorLogSink and
LogVectorOutput, respectively.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594)

### CODE REVIEW

**Review by rpoyner-tri — COMMENTED** [review]
+@seancurtis-tri has graciously agreed to feature review.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594#-:-MglqzX0EHdELtsYoBA3:bmsqkce)* status: 1 unresolved discussion, LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers (waiting on @SeanCurtis-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mglr3PeDMdyHKnzPyr0:-Mglr3PeDMdyHKnzPyr1:by57afr) (no related file):*
Unresolved anzu angst: this PR deprecates the old solution, and so motivates a desire to port to the new in downstream projects. I've done this exercise several times during work on the PR train, and reliably get test failures in one of the anzu test programs. Does that block this PR at all, or can we just roll and debug the migration problems separately?

A stale anzu migration example is at anzu::#7361. I will update it and see how it fares now.

---

**Comment by jwnimmer-tri** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mglr3PeDMdyHKnzPyr0:-MgltFY03Dv8E2Rx1Qqd:b-fmmmw6) (no related file):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

Unresolved anzu angst: this PR deprecates the old solution, and so motivates a desire to port to the new in downstream projects. I've done this exercise several times during work on the PR train, and reliably get test failures in one of the anzu test programs. Does that block this PR at all, or can we just roll and debug the migration problems separately?

A stale anzu migration example is at anzu::#7361. I will update it and see how it fares now.
</blockquote></details>

It's valid to work on this PR and an Anzu PR in parallel, and it's valid to land a Drake PR without perfecting a related Anzu PR as a precondition of the merge to Drake master.

However, we do need maintain the invariant that `bazel test //...` against Anzu master + Drake master is almost always passing, because using an unpinned Drake + Anzu master is an important, supported use case.  We do allow for the invariant to be broken for a few hours to allow for staggered merges, but if the Anzu fallout isn't fixed within those few hours, the response will be to revert the Drake PR, in order to restore the invariant in a timely manner.

---

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594)* status: 1 unresolved discussion, LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers (waiting on @jwnimmer-tri and @SeanCurtis-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mglr3PeDMdyHKnzPyr0:-Mglva1s-LXfDWT8Zl4f:b6lb0jy) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It's valid to work on this PR and an Anzu PR in parallel, and it's valid to land a Drake PR without perfecting a related Anzu PR as a precondition of the merge to Drake master.

However, we do need maintain the invariant that `bazel test //...` against Anzu master + Drake master is almost always passing, because using an unpinned Drake + Anzu master is an important, supported use case.  We do allow for the invariant to be broken for a few hours to allow for staggered merges, but if the Anzu fallout isn't fixed within those few hours, the response will be to revert the Drake PR, in order to restore the invariant in a timely manner.
</blockquote></details>

I don't think it will come to that as of this PR -- this just deprecates.

---

**Review by SeanCurtis-TRI — COMMENTED** [review]
Generally fine. The big issue is the breaking change to `pyplot_visualizer` (see below).

> Reviewed 31 of 31 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594#-:-Mglx4yA1vpYd-NTnDIq:b-z0c86l)* status: 7 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers (waiting on @jwnimmer-tri and @rpoyner-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mglr3PeDMdyHKnzPyr0:-MglxG8vB2Bj1bJTLio7:br1o6z6) (no related file):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

I don't think it will come to that as of this PR -- this just deprecates.
</blockquote></details>

@jwnimmer-tri I've got a related question. Can I do a PR with two commits: one updates Drake the second applies the patch? 

---
___
*[bindings/pydrake/systems/pyplot_visualizer.py, line 116 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15594#-MgltglpAlD4Z2sYDr0S:-MgltglpAlD4Z2sYDr0T:b-kmpfwe) ([raw file](https://github.com/robotlocomotion/drake/blob/4c723ea98a0c79b2e48991aa25b762ce92505d13/bindings/pydrake/systems/pyplot_visualizer.py#L116)):*
> ```Python
>         return ani
> 
>     def animate(self, log, resample=True, **kwargs):
> ```

nit: This is a breaking change and defies the deprecation. The user doesn't have the option to ignore deprecation warnings and resolve it some time in the future, their code will immediately be broken.

During the deprecation period, you should test to see if `log` is `VectorLog` or `SignalLogger` and act accordingly. Because you only deprecated the constructor, you can safely exercise the API without spewing additional warnings. And then the `SignalLogger` branch gets with the class.
___
*[bindings/pydrake/systems/test/primitives_test.py, line 116 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mglu70qCnDzCWH23q4s:-Mglu70qCnDzCWH23q4t:bkegaze) ([raw file](https://github.com/robotlocomotion/drake/blob/4c723ea98a0c79b2e48991aa25b762ce92505d13/bindings/pydrake/systems/test/primitives_test.py#L116)):*
> ```Python
>     @numpy_compare.check_nonsymbolic_types
>     def test_signal_logger(self, T):
>         with catch_drake_warnings(expected_count=3):
> ```

BTW This is a pretty big hammer. If anything else needs to get deprecated relative to this test in the next three months, it'll be a nuisance to figure out which of these calls *actually* triggered the warnings.
___
*[bindings/pydrake/systems/test/pyplot_visualizer_test.py, line 76 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15594#-MgluSxd5gz0eh4k6LTZ:-MgluSxd5gz0eh4k6LT_:b9tqdgz) ([raw file](https://github.com/robotlocomotion/drake/blob/4c723ea98a0c79b2e48991aa25b762ce92505d13/bindings/pydrake/systems/test/pyplot_visualizer_test.py#L76)):*
> ```Python
>         builder = DiagramBuilder()
>         system = builder.AddSystem(SimpleContinuousTimeSystem())
>         logger = builder.AddSystem(VectorLogSink(1))
> ```

nit Based on my comment about *not* breaking the `animate()` method, this test will need to preserve the `SignalLogger`-flavored version of this test.
___
*[systems/discrete_systems.h, line 86 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mglvy024OCiVX-sy-W5:-Mglvy024OCiVX-sy-W6:b-2x1ut9) ([raw file](https://github.com/robotlocomotion/drake/blob/4c723ea98a0c79b2e48991aa25b762ce92505d13/systems/discrete_systems.h#L86)):*
> ```objc
>   auto example = builder.AddSystem<ExampleDiscreteSystem>();
>   auto logger = LogVectorOutput(example->GetOutputPort("Sn"), &builder);
>   logger->SetPublishPeriod(ExampleDiscreteSystem::kPeriod);
> ```

nit: Is this method still a thing? I suspect not. I think you need to respell the example of calling `LogVectorOutput` to include the periodic event.
___
*[systems/primitives/signal_log.h, line 19 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15594#-MglwWNo2zoYICCKYDlf:-MglwWNpDfSFvuvJHOVH:bhe099t) ([raw file](https://github.com/robotlocomotion/drake/blob/4c723ea98a0c79b2e48991aa25b762ce92505d13/systems/primitives/signal_log.h#L19)):*
> ```objc
>  */
> template <typename T>
> class DRAKE_DEPRECATED("2021-11-01", "Use VectorLog instead.") SignalLog {
> ```

nit: This should be December.
___
*[systems/primitives/signal_logger.h, line 56 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mglw_P6Er3UkCYUAXlc:-Mglw_P6Er3UkCYUAXld:b-3yldha) ([raw file](https://github.com/robotlocomotion/drake/blob/4c723ea98a0c79b2e48991aa25b762ce92505d13/systems/primitives/signal_logger.h#L56)):*
> ```objc
> /// @ingroup primitive_systems
> template <typename T>
> class DRAKE_DEPRECATED("2021-11-01", "Use VectorLogSink instead.")
> ```

nit: December (multiple times).
___
*[systems/primitives/test/discrete_derivative_test.cc, line 53 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15594#-MglwlEgA7TADGspn1gy:-MglwlEgA7TADGspn1gz:bg7nt80) ([raw file](https://github.com/robotlocomotion/drake/blob/4c723ea98a0c79b2e48991aa25b762ce92505d13/systems/primitives/test/discrete_derivative_test.cc#L53)):*
> ```cpp
>   builder.Connect(ppsource->get_output_port(), deriv->get_input_port());
>   // Evaluate the outputs at a distinct sampling interval.
>   auto sink = LogVectorOutput(deriv->get_output_port(), &builder,
> ```

BTW I've seen a lot of "logger" in this PR, but I think this is the first "sink" I've seen. Intentional?

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594)* status: 7 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers (waiting on @rpoyner-tri and @SeanCurtis-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mglr3PeDMdyHKnzPyr0:-MglxUxt418mfAn8Nq2x:b-ai5k01) (no related file):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

@jwnimmer-tri I've got a related question. Can I do a PR with two commits: one updates Drake the second applies the patch? 
</blockquote></details>

@rpoyner-tri Aha, my answer was incomplete.  Another invariant is that we never spam the console (e.g., with deprecation warnings) during a build, or when running a python program on the command line.  However, that doesn't imply that you need to do the Anzu porting and hunt down the bug immediately.  Instead, the Anzu PR you'd need to submit ASAP is one that muted the deprecation warnings.

@SeanCurtis-TRI Yes, that is a supported workflow, and in fact is the only way possible when Drake has certain kinds of breaking changes.  When a Drake upgrade does _not_ have breaking changes, though, I try to encourage people to first do the sha bump, merge that, and then submit the follow-up separately.  That _proves_ that there we no breaking changes.

---

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594)* status: 7 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers (waiting on @jwnimmer-tri and @rpoyner-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mglr3PeDMdyHKnzPyr0:-MglzoGJ-JVfimLT1IyA:b-senttq) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

@rpoyner-tri Aha, my answer was incomplete.  Another invariant is that we never spam the console (e.g., with deprecation warnings) during a build, or when running a python program on the command line.  However, that doesn't imply that you need to do the Anzu porting and hunt down the bug immediately.  Instead, the Anzu PR you'd need to submit ASAP is one that muted the deprecation warnings.

@SeanCurtis-TRI Yes, that is a supported workflow, and in fact is the only way possible when Drake has certain kinds of breaking changes.  When a Drake upgrade does _not_ have breaking changes, though, I try to encourage people to first do the sha bump, merge that, and then submit the follow-up separately.  That _proves_ that there we no breaking changes.
</blockquote></details>

Thanks (I have some minor clean up related to #15534).

---

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594)* status: 6 unresolved discussions, LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers (waiting on @rpoyner-tri)

___
*[systems/discrete_systems.h, line 86 at r1](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mglvy024OCiVX-sy-W5:-Mgm10359FEq8WgkZU3V:bwopym4) ([raw file](https://github.com/robotlocomotion/drake/blob/4c723ea98a0c79b2e48991aa25b762ce92505d13/systems/discrete_systems.h#L86)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

nit: Is this method still a thing? I suspect not. I think you need to respell the example of calling `LogVectorOutput` to include the periodic event.
</blockquote></details>

yup, looks stale and busted.

**Review by rpoyner-tri — COMMENTED** [review]
+(status: do not merge) holding until I work out an accompanying Anzu PR. Other review can continue.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594#-:-MgmVcJp9VErDRsG-SDE:b-tq4ppb)* status: LGTM missing from assignee SeanCurtis-TRI(platform), needs at least two assigned reviewers, labeled "do not merge" (waiting on @SeanCurtis-TRI)

**Review by SeanCurtis-TRI — COMMENTED** [review]
:LGTM: +@sammy-tri for platform review, as per schedule.

> Reviewed 8 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594#-:-MgpauvAD8XwdQ197L4n:bh3myp9)* status: 1 unresolved discussion, LGTM missing from assignee sammy-tri(platform), labeled "do not merge" (waiting on @rpoyner-tri and @sammy-tri)

___
*[bindings/pydrake/systems/pyplot_visualizer.py, line 135 at r2](https://reviewable.io/reviews/robotlocomotion/drake/15594#-MgpaUMPETpUSojmrPlc:-MgpaUMPETpUSojmrPld:bxfyk4k) ([raw file](https://github.com/robotlocomotion/drake/blob/4255ccc34d1c93b13cb320433cc29088679bf4ed/bindings/pydrake/systems/pyplot_visualizer.py#L135)):*
> ```Python
>         if log_is_signal_logger and not self._warned_signal_logger_deprecated:
>             _warn_deprecated("SignalLogger is deprecated. Use VectorLog and"
>                              " VectorLogSink instead.", date="2021-12-01")
> ```

nit: In this *particular* case, I believe reference to `VectorLogSink` is misleading -- that would *not* be a valid parameter type.

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594)* status: LGTM missing from assignee sammy-tri(platform), labeled "do not merge" (waiting on @sammy-tri and @SeanCurtis-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mglr3PeDMdyHKnzPyr0:-MgpnAljD6YTuWMtUNm6:b-tixn0r) (no related file):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Thanks (I have some minor clean up related to #15534).
</blockquote></details>

anzu::#7361 is behaving better -- not quite perfect yet.

---

**Review by SeanCurtis-TRI — COMMENTED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594)* status: LGTM missing from assignee sammy-tri(platform), labeled "do not merge" (waiting on @sammy-tri)

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594)* status: 1 unresolved discussion, LGTM missing from assignee sammy-tri(platform), labeled "do not merge" (waiting on @rpoyner-tri and @sammy-tri)

___
*[examples/acrobot/spong_sim.cc, line 123 at r4](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mgq0YyO5f2aFgBkJbHO:-Mgq0YyO5f2aFgBkJbHP:b-buntaz) ([raw file](https://github.com/robotlocomotion/drake/blob/8208bd7e317b0dda462bdc551833c31ca5ec9b3c/examples/acrobot/spong_sim.cc#L123)):*
> ```cpp
> 
>   Output output;
>   output.x_tape = state_logger->FindLog(simulator.get_context()).data();
> ```

one last bit of fiddling here to minimize divergence with the twin file in anzu.

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594)* status: LGTM missing from assignee sammy-tri(platform), labeled "do not merge" (waiting on @sammy-tri and @SeanCurtis-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mglr3PeDMdyHKnzPyr0:-Mgq3pIZDirDkh8EgoM1:bfi85ea) (no related file):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

anzu::#7361 is behaving better -- not quite perfect yet.
</blockquote></details>

The anzu patch is ready to go; just awaiting drake bump and various CI cycles. I'm content to hold this patch until the anzu::#7361 is merged.

---

**Review by rpoyner-tri — COMMENTED** [review]
-(status: do not merge) anzu patch is merged.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594#-:-MgqGhZCEJFy2vmDXgMI:b-gaycmh)* status: LGTM missing from assignee sammy-tri(platform), labeled "do not merge" (waiting on @sammy-tri)

**Review by sammy-tri — COMMENTED** [review]
> Reviewed 22 of 31 files at r1, 7 of 8 files at r2, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594)* status: 1 unresolved discussion, LGTM missing from assignee sammy-tri(platform) (waiting on @rpoyner-tri)

___
*[systems/discrete_systems.h, line 95 at r5](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mgqt2_e5mpsfxQug0ec:-Mgqt2_e5mpsfxQug0ed:b-zbf6hw) ([raw file](https://github.com/robotlocomotion/drake/blob/cb909203fcb243765f01e60bbad5be3c8e93a59d/systems/discrete_systems.h#L95)):*
> ```objc
>   // Print out the contents of the log.
>   const auto& log = logger.FindLog(simulator.get_context());
>   for (int n = 0; n < log->sample_times().size(); ++n) {
> ```

Should the `log->` examples in this comment now read `log.`?

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594)* status: 1 unresolved discussion, LGTM missing from assignee sammy-tri(platform) (waiting on @rpoyner-tri)

___
*[systems/discrete_systems.h, line 95 at r5](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mgqt2_e5mpsfxQug0ec:-MgqvayiAOYT-MQTivAq:b-kl8g3z) ([raw file](https://github.com/robotlocomotion/drake/blob/cb909203fcb243765f01e60bbad5be3c8e93a59d/systems/discrete_systems.h#L95)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Should the `log->` examples in this comment now read `log.`?
</blockquote></details>

good catch.

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/15594)* status: 1 unresolved discussion, LGTM missing from assignee sammy-tri(platform) (waiting on @sammy-tri and @SeanCurtis-TRI)

___
*[systems/discrete_systems.h, line 95 at r5](https://reviewable.io/reviews/robotlocomotion/drake/15594#-Mgqt2_e5mpsfxQug0ec:-Mgqx5Be3PlAF6dSoEJe:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/cb909203fcb243765f01e60bbad5be3c8e93a59d/systems/discrete_systems.h#L95)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

good catch.
</blockquote></details>

Done.

File: examples/manipulation_station/end_effector_teleop_sliders.py [hunk 1]
```diff
@@ -339,14 +339,15 @@ def main():
...
-        iiwa_velocities = builder.AddSystem(SignalLogger(num_iiwa_joints))
+        iiwa_velocities = builder.AddSystem(VectorLogSink(num_iiwa_joints))
         builder.Connect(station.GetOutputPort("iiwa_velocity_estimated"),
                         iiwa_velocities.get_input_port(0))
     else:
         iiwa_velocities = None
 
     diagram = builder.Build()
     simulator = Simulator(diagram)
+    iiwa_velocities_log = iiwa_velocities.FindLog(simulator.get_context())
```
Comment (jwnimmer-tri) [inline examples/manipulation_station/end_effector_teleop_sliders.py:350]: This line crashes (cannot call FindLog on None) when run in non-test mode.

**Pull request merged**